numpy/lib/function_base.py

```diff
@@ -189,40 +199,86 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
         if mn == mx:
             mn -= 0.5
             mx += 0.5
+        # At this point, if the weights are not integer, floating point, or
+        # complex, we have to use the slow algorithm.
```

Comment:
What is the rationale for this? What type of weight would not work here?

Reply:
I can't remember for sure, but I think e.g. object arrays work in the weights array:

```
In [7]: np.histogram([1,2,3], weights=np.array([Decimal(1), Decimal(2.2), Decimal(4.4)], dtype=np.object))
Out[7]: 
(array([Decimal('1'), Decimal('0'), Decimal('0'), Decimal('0'),
        Decimal('0'), Decimal('2.200000000000000177635683940'),
        Decimal('0E-27'), Decimal('0E-27'), Decimal('0E-27'),
        Decimal('4.400000000000000355271367880')], dtype=object),
 array([ 1. ,  1.2,  1.4,  1.6,  1.8,  2. ,  2.2,  2.4,  2.6,  2.8,  3. ]))
```

Reply:
Yep, that's it, thanks - you're guarding `bincount`, that only works for the types you list there.